File: cpp/base/src/main/vsdk/toolkit/environment/geometry/geometricProcessing/polygonClipper/_Polygon2DWA.cpp
#include <cmath>

#include "java/util/ArrayList.txx"
#include "vsdk/toolkit/environment/geometry/geometricProcessing/polygonClipper/_Polygon2DWA.h"
_Polygon2DWA::_Polygon2DWA() : currentLoop(0) { nextLoop(); }

_Polygon2DWA::_Polygon2DWA(const Polygon2D& polyToCopy, bool copyClean) : currentLoop(0)
{
    bool oneOrMoreLoops = false;
    for (long int i = 0; i < polyToCopy.loops.size(); ++i) {
        _Polygon2DContour* contourToCopy = polyToCopy.loops.get(i);
        oneOrMoreLoops = true;
        nextLoop();
        if (contourToCopy->vertices.size() == 0) continue;
        if (contourToCopy->vertices.size() == 1) {
            Vertex2D prevV = contourToCopy->vertices.get(0);
            addVertex(prevV.x, prevV.y, prevV.color.r(), prevV.color.g(), prevV.color.b());
        }
        else {
            Vertex2D prevV = contourToCopy->vertices.get(contourToCopy->vertices.size()-1);
            Vertex2D prevPV = contourToCopy->vertices.get(contourToCopy->vertices.size()-2);
            bool removeLast = false;
            bool isFirst = true;
            for (long int j = 0; j < contourToCopy->vertices.size(); ++j) {
                Vertex2D v = contourToCopy->vertices.get(j);
                if (copyClean) {
                    if (std::fabs(prevV.x-v.x) > 0.0001 || std::fabs(prevV.y-v.y) > 0.0001) {
                        addVertex(v.x, v.y, v.color.r(), v.color.g(), v.color.b());
                        if (areCollinearAndOposite2DVectors(prevPV, prevV, v)) {
                            if (isFirst) removeLast = true;
                            else currentLoop->removeVertex(currentLoop->vertices.size()-2);
                        }
                        else prevPV = prevV;
                        prevV = v;
                    }
                    isFirst = false;
                }
                else addVertex(v.x, v.y, v.color.r(), v.color.g(), v.color.b());
            }
            if (removeLast) currentLoop->removeVertex(currentLoop->vertices.size()-1);
            if (currentLoop->vertices.size() == 0) {
                Vertex2D p = contourToCopy->vertices.get(0);
                addVertex(p.x, p.y, p.color.r(), p.color.g(), p.color.b());
            }
        }
    }
    if (!oneOrMoreLoops) nextLoop();
}

_Polygon2DWA::~_Polygon2DWA()
{
    for (long int i = 0; i < loops.size(); ++i) delete loops[i];
}

bool _Polygon2DWA::areCollinearAndOposite2DVectors(const Vertex2D& vEndA, const Vertex2D& vStartAB, const Vertex2D& vEndB)
{
    Vertex2D a(vEndA.x - vStartAB.x, vEndA.y - vStartAB.y);
    Vertex2D b(vEndB.x - vStartAB.x, vEndB.y - vStartAB.y);
    double t = std::sqrt(a.x*a.x + a.y*a.y);
    if (t < 1e-12) return false;
    a.x /= t; a.y /= t;
    t = std::sqrt(b.x*b.x + b.y*b.y);
    if (t < 1e-12) return false;
    b.x /= t; b.y /= t;
    t = a.x*b.x + a.y*b.y;
    return (t > -1.0001 && t < -0.9999);
}

void _Polygon2DWA::addVertex(double x, double y, double r, double g, double b) { currentLoop->addVertex(x,y,r,g,b); }
void _Polygon2DWA::addVertex(double x, double y) { currentLoop->addVertex(x,y); }
void _Polygon2DWA::pushVertex(double x, double y) { currentLoop->pushVertex(x,y); }
void _Polygon2DWA::nextLoop() { currentLoop = new _Polygon2DContourWA(); loops.add(currentLoop); }

File: typescript/base/src/main/vsdk/toolkit/environment/geometry/geometricProcessing/polygonClipper/_Polygon2DWA.ts
import { Vertex2D } from "../../element/Vertex2D.js";
import { Polygon2D } from "../../surface/polygon/Polygon2D.js";
import { _Polygon2DContourWA } from "./_Polygon2DContourWA.js";

/** Polygon representation consumed by the Weiler--Atherton implementation. */
export class _Polygon2DWA {
    public loops: _Polygon2DContourWA[] = [];
    private currentLoop!: _Polygon2DContourWA;

    public constructor();
    public constructor(polyToCopy: Polygon2D, copyClean: boolean);
    public constructor(polyToCopy?: Polygon2D, copyClean = false) {
        if (polyToCopy === undefined) {
            this.nextLoop();
            return;
        }
        let oneOrMoreLoops = false;
        for (const contour of polyToCopy.loops) {
            oneOrMoreLoops = true;
            this.nextLoop();
            if (contour.vertices.length === 0) continue;
            if (contour.vertices.length === 1) {
                const vertex = contour.vertices[0]!;
                this.addVertex(vertex.x, vertex.y, vertex.color.r(), vertex.color.g(), vertex.color.b());
                continue;
            }
            let previousVertex = contour.vertices[contour.vertices.length - 1]!;
            let previousPreviousVertex = contour.vertices[contour.vertices.length - 2]!;
            let removeLast = false;
            let isFirst = true;
            for (const vertex of contour.vertices) {
                if (!copyClean) {
                    this.addVertex(vertex.x, vertex.y, vertex.color.r(), vertex.color.g(), vertex.color.b());
                } else if (
                    Math.abs(previousVertex.x - vertex.x) > 0.0001 ||
                    Math.abs(previousVertex.y - vertex.y) > 0.0001
                ) {
                    this.addVertex(vertex.x, vertex.y, vertex.color.r(), vertex.color.g(), vertex.color.b());
                    if (this.areCollinearAndOposite2DVectors(previousPreviousVertex, previousVertex, vertex)) {
                        if (isFirst) removeLast = true;
                        else this.currentLoop.removeVertex(this.currentLoop.vertices.size() - 2);
                    } else previousPreviousVertex = previousVertex;
                    previousVertex = vertex;
                }
                isFirst = false;
            }
            if (removeLast) this.currentLoop.removeVertex(this.currentLoop.vertices.size() - 1);
            if (this.currentLoop.vertices.size() === 0) {
                const vertex = contour.vertices[0]!;
                this.addVertex(vertex.x, vertex.y, vertex.color.r(), vertex.color.g(), vertex.color.b());
            }
        }
        if (!oneOrMoreLoops) this.nextLoop();
    }

    public addVertex(x: number, y: number, r?: number, g?: number, b?: number): void {
        this.currentLoop.addVertex(x, y, r, g, b);
    }
    public pushVertex(x: number, y: number): void {
        this.currentLoop.pushVertex(x, y);
    }
    public nextLoop(): void {
        this.currentLoop = new _Polygon2DContourWA();
        this.loops.push(this.currentLoop);
    }

    private areCollinearAndOposite2DVectors(endA: Vertex2D, startAB: Vertex2D, endB: Vertex2D): boolean {
        let ax = endA.x - startAB.x;
        let ay = endA.y - startAB.y;
        let bx = endB.x - startAB.x;
        let by = endB.y - startAB.y;
        const aLength = Math.hypot(ax, ay);
        const bLength = Math.hypot(bx, by);
        ax /= aLength;
        ay /= aLength;
        bx /= bLength;
        by /= bLength;
        const dot = ax * bx + ay * by;
        return dot > -1.0001 && dot < -0.9999;
    }
}
